### PR TITLE
Halo add pb test for halo exchange

### PR DESCRIPTION
**Pull request by TomMelt**
# Add periodic boundary test for Halo Exchange

### Task List
- [x] Linked an issue above that captures the requirements of this PR
- [ ] Defined the tests that specify a complete and functioning change
- [ ] Implemented the source code change that satisfies the tests
- [ ] Commented all code so that it can be understood without additional context
- [ ] No new warnings are generated or they are mentioned below
- [ ] The documentation has been updated (or an issue has been created to do so)
- [x] Relevant labels (e.g., enhancement, bug) have been applied to this PR
- [x] This change conforms to the conventions described in the README

---
# Change Description

- Adds a test for halo exchange with periodic boundaries

---
# Test Description

### `HaloExchangePB_test.cpp`

`HaloExchangePB_test.cpp` is based on `HaloExchangeCB_test.cpp`. It creates test data for all the array types:
- `HField`
- `VertexField`
- `DGField`
- `DGVector`
- `CGVector`

They  are created in such a way that the value of each cell can be calculated based on its indices. Therefore we can check after exchange that all of the cells have been exchanged successfully.

This would be a good PR for @niravshah241 to take over.

### CODE REVIEW

**Comment by Mikolaj-A-Kowalski** [comment]
Resurrecting and taking over this PR. 

To get started rebased on the current `develop`.

**Comment by Mikolaj-A-Kowalski** [comment]
This one is now ready to go in my opinion. 

The main things to discuss are 
- How clear is the wraparound logic and the mapping `local` -> `global` for the cells and points coordinates? I found it to be the most confusing part of the `CB` version of the test and have changed it significantly in the `PB` version. 
- Should we refactor `CB` test as well to work with signed integers as coordinates?

**Comment by Mikolaj-A-Kowalski** [comment]
Ok. We are running into issues with the failing CI but these are related by some issues with the `entrypoint.sh` scripts.

**Comment by Mikolaj-A-Kowalski** [comment]
With the Docker image being back online I think this one is ready for review now.

**Comment by niravshah241** [comment]
I tried to run the tests ``testHaloExchangeCB_MPI3`` and ``testHaloExchangePB_MPI3`` an both are failing. Is this PR ready for review? I also think that the task list is probably done but not marked with a check sign.

**Comment by Mikolaj-A-Kowalski** [comment]
> I tried to run the tests testHaloExchangeCB_MPI3 and testHaloExchangePB_MPI3 an both are failing. Is this PR ready for review? I also think that the task list is probably done but not marked with a check sign.

It is ready for review and the tests were working fine on my side (CI also seems to be happy). What kind of error message are you getting?

File: core/test/HaloExchangePB_test.cpp [hunk 1]
```diff
@@ -0,0 +1,415 @@
...
+ * @brief   Test halo exchange class for Periodic Boundaries (PB)
+ * @details
+ *
+ * The purpose of this test is to verify the Halo exchange.
+ *
+ * We accomplish that by:
+ *  - Assigning each value in the grid with a unique id (hash) based on its global
+ *    coordinates
+ *  - We take care to not assign any values in the halo region for each rank
+ *  - We then perform tha halo exchange
```
Comment (jatkinson1000) [inline core/test/HaloExchangePB_test.cpp:12]:
```suggestion
 *  - We then perform the halo exchange
```

File: core/test/HaloExchangePB_test.cpp [hunk 2]
```diff
@@ -0,0 +1,415 @@
...
+ *  - For each rank we verify that the values stored in the field match the
+ *    hash
```
Comment (jatkinson1000) [inline core/test/HaloExchangePB_test.cpp:14]: Not clear what this means.
You check that the hash recomputed using the remapped variables matches the original? You check that the values at the new location match those at the origin hash? You check that the hash recomputed at the new location matches the origin and that the values are the same?

File: core/test/HaloExchangePB_test.cpp [hunk 3]
```diff
@@ -0,0 +1,415 @@
...
+static const int CGDEGREE = 2;
+static const bool debug = false;
+
+/**
+ * Helper class to group all domain-dependent calculations
+ *
+ * It is hardcoded for a periodic topology both in X and Y.
+ * The interpretation of  wraparound for the cell based fields (e.g. H or DG) is
+ * quite straightforward. It is just that on the right edge, the right neighbours
+ * of the cells are the cells on the left edge of the domain.
```
Comment (jatkinson1000) [inline core/test/HaloExchangePB_test.cpp:51]: Consider a diagram to be explicit?

```suggestion
 * of the cells are the cells on the left edge of the domain.
 * 
 * | 0 | 1 | 2 | . . . | n-1 | n | n+1 = 0
```

Also you state that this is in X and Y, so also upper neighbours of the top cells are the bottom cells in the domain? Maybe just be explicit about this too?

File: core/test/HaloExchangePB_test.cpp [hunk 4]
```diff
@@ -0,0 +1,415 @@
...
+ *
+ * For point based fields there is only one caveat. The points on the right edge
+ * ARE the points on the left edge. There is continuity and they need to have
+ * the same identity.
+ *
```
Comment (jatkinson1000) [inline core/test/HaloExchangePB_test.cpp:56]: I don't understand this.

Do you mean fields that are defined on vertices or edges?

To me this does not need saying as it is already included in the above definition and this just serves to confuse, but perhaps that's just me.

If you want you could just leave as "Note: that this means that the vertices and edges on the left boundary are identical to those on the right."

Comment (jatkinson1000) [reply core/test/HaloExchangePB_test.cpp:56]: Ah, it relates to the below statement about number of cells. I think I would wrap this up in a "Note" rather than calling it a caveat. To me it is not a caveat, but follows from what you have already defined and is just a nuance to be aware of.

File: core/test/HaloExchangePB_test.cpp [hunk 5]
```diff
@@ -0,0 +1,415 @@
...
+        , ny(ny)
+        , localNx(localNx)
+        , localNy(localNy)
+        , offsetX(offsetX)
+        , offsetY(offsetY)
+        , haloSize(1)
+        , type(type)
+    {
+        if (type == ModelArray::Type::VERTEX) {
+            //  Vertex fields have an extra entries (e.g. 3 cells in 1D have 4 vertices)
```
Comment (jatkinson1000) [inline core/test/HaloExchangePB_test.cpp:95]: Ie, not periodic, per the above description? Maybe reinforce that?

File: core/test/HaloExchangePB_test.cpp [hunk 6]
```diff
@@ -0,0 +1,415 @@
...
+                offset *= 10;
+            }
+            return offset;
+        }(domain_size);
+
+        return component_offset * (component + 1) + globalI * this->ny + globalJ;
+    }
+
+    /**
+     * @brief Map the local coorinates of the domain to the global coordinates
```
Comment (jatkinson1000) [inline core/test/HaloExchangePB_test.cpp:138]:
```suggestion
     * @brief Map the local coordinates of the domain to the global coordinates
```

File: core/test/HaloExchangePB_test.cpp [hunk 7]
```diff
@@ -0,0 +1,415 @@
...
+    verifyTestData(testData, DGVectorDomain);
+}
+
+MPI_TEST_CASE("CGVector", 3)
+{
+    auto& modelMPI = ModelMPI::getInstance();
+    auto& metadata = ModelMetadata::getInstance();
+
+    // We work on signed integers to represent extents and grid
+    // coordinates
```
Comment (jatkinson1000) [inline core/test/HaloExchangePB_test.cpp:367]: Is this significant? Different to what neXtSIM usually does?

File: core/test/HaloExchangePB_test.cpp (hunk 4 above)
Comment (jatkinson1000) [reply core/test/HaloExchangePB_test.cpp:56]: From discussion:

For subdomains points on the left/bottom edge belong to the domain, and points on the right/top belong to the right/top-neighbouring domain.

For global periodic wrap-around means left/bottom edge is also the right/top.

But for non-periodic a separate top/right set of values have to be defined to set the boundary condition.

File: core/test/HaloExchangePB_test.cpp (hunk 5 above)
Comment (jatkinson1000) [reply core/test/HaloExchangePB_test.cpp:95]: From discussion, comment is perhaps stale and needs clarification for local subdomain which always includes halos, so is always n+1 vertices for n cells.

File: core/test/HaloExchangePB_test.cpp [hunk 8]
```diff
@@ -0,0 +1,415 @@
...
+ *  - We then perform tha halo exchange
+ *  - For each rank we verify that the values stored in the field match the
+ *    hash
+ *
+ * Test halo exchange class for Periodic Boundaries (PB) on following fields:
+ * - HField
+ * - VertexField
+ * - DGField
+ * - DGVector
+ * - CGVector
```
Comment (jatkinson1000) [inline core/test/HaloExchangePB_test.cpp:21]: Clarify here or in test suite that the same test (3 subdomains) is run on each of these types?

**Review by jatkinson1000 — COMMENTED** [review]
Hey @Mikolaj-A-Kowalski generally looking pretty good.

A couple of typos and some suggestions for clarifying the explanation (to me at least).

File: core/test/HaloExchangePB_test.cpp (hunk 7 above)
Comment (Mikolaj-A-Kowalski) [reply core/test/HaloExchangePB_test.cpp]: It is different with respect to the sibling CB version of the test that works on the unsigned integers, which I am feeling makes the global position calculations more difficult to perform. 

I will expand the comment to elaborate this.

File: core/test/HaloExchangePB_test.cpp (hunk 2 above)
Comment (Mikolaj-A-Kowalski) [reply core/test/HaloExchangePB_test.cpp]: I will try to reword it. The gist that I was trying to get across is that: 
- The hash value is dependent only on the position of the cell in the global coordinates and the component index
- We store the value of the hash in the field itself. But do not assign values in the halo regions
- We perform halo exchange 
- We can now go through each rank and compare the value in the field against the hash. If all matches we know that the exchange was correct and all values transferred to their correct location.

Comment (Mikolaj-A-Kowalski) [reply core/test/HaloExchangePB_test.cpp]: Changed in e8c8a8cef78c

File: core/test/HaloExchangePB_test.cpp (hunk 8 above)
Comment (Mikolaj-A-Kowalski) [reply core/test/HaloExchangePB_test.cpp]: Changed in  e8c8a8cef78

File: core/test/HaloExchangePB_test.cpp (hunk 5 above)
Comment (Mikolaj-A-Kowalski) [reply core/test/HaloExchangePB_test.cpp]: Expanded this remark in 0b81fdccf7

File: core/test/HaloExchangePB_test.cpp (hunk 7 above)
Comment (Mikolaj-A-Kowalski) [reply core/test/HaloExchangePB_test.cpp]: Expanded in 7efcec3316

File: core/test/HaloExchangePB_test.cpp (hunk 3 above)
Comment (Mikolaj-A-Kowalski) [reply core/test/HaloExchangePB_test.cpp:56]: I have added the diagrams in 50f8dc7a7e68d04

File: core/test/HaloExchangePB_test.cpp (hunk 4 above)
Comment (Mikolaj-A-Kowalski) [reply core/test/HaloExchangePB_test.cpp]: I have tried to clarify these points in the expanded discussion in 50f8dc7a7e68d04 

Hopefully they become much cleared without the unnecessary new paragraph (that confused you) and more visual illustration with the diagrams.

**Comment by Mikolaj-A-Kowalski** [comment]
@jatkinson1000 I have tried to address all of your comments. Please mark the threads as resolved if the changes fixes the points to your satisfaction ;-)

File: core/test/HaloExchangePB_test.cpp (hunk 2 above)
Comment (jatkinson1000) [reply core/test/HaloExchangePB_test.cpp]: I think that reads better, thanks. One last point to clarify - when verifying you do so by recomputing the hash based on global coordinate and checking it matches the value that was transferred during halo exchange? This last step wasn't explicitly clear.

File: core/test/HaloExchangePB_test.cpp (hunk 4 above)
Comment (jatkinson1000) [reply core/test/HaloExchangePB_test.cpp]: Yes, diagrams are great, thanks!

**Review by jatkinson1000 — APPROVED** [review]
Thanks @Mikolaj-A-Kowalski one last query, but I think this is good to go.

File: core/test/HaloExchangePB_test.cpp (hunk 2 above)
Comment (Mikolaj-A-Kowalski) [reply core/test/HaloExchangePB_test.cpp]: > One last point to clarify - when verifying you do so by recomputing the hash based on global coordinate and checking it matches the value that was transferred during halo exchange?

Exactly. I think the point that is not 100% clear is that we do it for all values in the field (halo and exterior).  This is to detect cases when e.g. we mess up the indexing and corrupt data in the interior. 

Will expand it in a sec.

Comment (Mikolaj-A-Kowalski) [reply core/test/HaloExchangePB_test.cpp]: Added the motivation for checking the whole field.

**Comment by Mikolaj-A-Kowalski** [comment]
> Thanks @Mikolaj-A-Kowalski one last query, but I think this is good to go.

@jatkinson1000  Reworded the remark one more time ;-) 
Also rebased on the current develop. If you are happy I will proceed and merge when the tests pass.

**Comment by jatkinson1000** [comment]
Gah, fixups 😛 

LGTM, you may merge when ready.
